Update Mixtrack 3 mapping to support #1187

radusuciu opened the pull request:
Just bringing up the mapping to support new controls introduced in #1187.  May take opportunity to re-map as a few button combinations have been freed up.

radusuciu committed:
Update Mixtrack 3 mapping to support #1187

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -1907,41 +1875,19 @@ NumarkMixtrack3.BeatKnob = function(channel, control, value, status, group) {
-    beatJumpSize = parseInt(control.match(/\d+/)[0]);
-};
+        if (increase) {
+            jumpSize = presetJumpSizes[(index + 1) % (presetJumpSizes.length)];
```
Is there a need for wrapping around when reaching the smallest/biggest value? When I was developing #1187 I found such behavior to be a little unexpected. I think checking against a minimum and maximum value is sufficient. This would also allow the user to enter a non-power-of-2 number into the spinbox and halve/double it from the controller.

radusuciu replied:
No need and I agree.

Be-ing reviewed (commented):
Thanks for updating this.

I suggest switching around the mapping so shift + pitch up/down doubles/halves beatjump_size and shift + beats encoder does the jumping. This would allow for quickly beatjumping through the track with repeated beatjumps. Using buttons for this would require repeatedly pressing the buttons, but it is easier to repeatedly turn an encoder.

Have you considered remapping the autoloop pad mode like the [Pioneer DDJ-SB2](https://mixxx.org/wiki/doku.php/pioneer_ddj-sb2#pads_-_auto_loop_mode)? Considering you already have other controls for beatjump_forward/backward, I think it would make more sense to use shift + halve/double as beatjump_1_forward/backward (independent from beatjump_size).

radusuciu commented:
> I suggest switching around the mapping so shift + pitch up/down doubles/halves beatjump_size and shift + beats encoder does the jumping. This would allow for quickly beatjumping through the track with repeated beatjumps. Using buttons for this would require repeatedly pressing the buttons, but it is easier to repeatedly turn an encoder.

Yes, I saw this, will have to play with it. Ideally I'd like encoders for both - which could be sort of accomplished by doing `shift + beats encoder` to change size and `tap + beats encoder` to actually jump.

> Have you considered remapping the autoloop pad mode like the Pioneer DDJ-SB2? Considering you already have other controls for beatjump_forward/backward, I think it would make more sense to use shift + halve/double as beatjump_1_forward/backward (independent from beatjump_size).

I enjoy the current autoloop setup on this controller but will consider changing the shift actions.

radusuciu committed:
Use min/max jump sizes instead of presets

Be-ing commented:
> Yes, I saw this, will have to play with it. Ideally I'd like encoders for both - which could be sort of accomplished by doing shift + beats encoder to change size and tap + beats encoder to actually jump.

It might get confusing if two different modifiers with the same encoder do slightly different functions.

> I enjoy the current autoloop setup on this controller but will consider changing the shift actions.

Then you would lose access to small loops on the controller. A benefit of the new loop controls are that they allow controllers to fit more functionality into the same physical components while allowing the user to access a wide range of loop sizes.

radusuciu committed:
Remove unused method

radusuciu committed:
Rename TrackIsLoaded method

Also using this method instead of checking `deck.duration`.

radusuciu committed:
Remove extraneous method

radusuciu committed:
Use new library controls

radusuciu committed:
Move away from manually keeping track of loaded

radusuciu committed:
Map sampler gain to pad + beats knob

radusuciu committed:
Move button init to constructors

radusuciu committed:
Move soft-takeover handling to constructors

radusuciu committed:
Partly move LED init to constructor

radusuciu committed:
Move deck and sampler init to init method

radusuciu committed:
Finish moving LEDs to appropriate constructors

radusuciu committed:
Samplers only have one LED to track

radusuciu committed:
Partial revert to deprecated library controls

The old library controls are better for my use case, and I imagine for
many people with smaller controllers. Suggestions on how to make use of
the new library controls are welcome..

radusuciu committed:
Remove extra comments

radusuciu committed:
Small refactor

radusuciu committed:
fix error when first switching to autoloop

radusuciu committed:
Simplify sampler gain handling

radusuciu committed:
Init controller as an object

radusuciu committed:
Show super knobs by default

When no effect is focused and instantFX are off, the touch strip is
mapped to control the super knob, thus it makes sense to have it be
visible.

radusuciu commented:
I think I'm about done with making changes for now. Did a bit of code clean-up, fixed a small but annoying bug and mostly kept the mapping the same. I thought about your suggestions but my personal preference is to keep things the way they are. I'm not opposed to re-visiting as the release of 2.1 gets closer.

I really like the new interface for looping and beatjumping by the way, nicely done!

Wiki has been updated to reflect these latest changes. Unless you've got additional suggestions, this is ready to merge as far as I'm concerned.

Thanks!

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -1887,6 +1843,54 @@ NumarkMixtrack3.BeatKnob = function(channel, control, value, status, group) {
+            var samplerGroup = '[Sampler' + samplersToToggle[i] + ']'
+            var gainValue = engine.getValue(samplerGroup, 'pregain');
+            var gainMultiplier = 3;
+            var maxGain = 4;
```
This relies on an implementation detail of the pregain ControlObject. How about using engine.set/getParameter instead?

radusuciu replied:
I'm a little unclear on the difference between `getValue` and `getParameter` tbh. Basically parameters perform scaling and bound checking between 0 and 1?

> “Parameter” values range from 0 to 1. If you set a parameter to 1 then the value of the control becomes the maximum value. If you set the parameter to 0 then the value of the control becomes the minimum value.

So basically `setParameter` and `setValue` are almost always preferred?

Be-ing replied:
set/getParameter are preferred for ControlObjects with a continuous range (knobs and faders) whereas set/getValue are preferred for ControlObjects with discrete values (binary or multi-state buttons). setParameter and setValue both perform bounds checking unless the ControlObject specifically allows setting outside its normal range (as mentioned above, I think only "rate" does this currently).

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -766,13 +770,16 @@ NumarkMixtrack3.deck.prototype.focusNextEffect = function() {
 
     this.PADSampleButtonHold = new LongShortBtn(NumarkMixtrack3.onPADSampleButtonHold);
+
+    engine.softTakeover(this.group, 'pregain', true);
```
This is manipulated by a relative control (Beats encoder), so soft takeover doesn't matter.

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -1799,31 +1799,13 @@ NumarkMixtrack3.BeatKnob = function(channel, control, value, status, group) {
-
             gainValue = gainValue + increment;
 
             if (gainValue < 0) {
```
ControlObjects take care of value clamping automatically unless they specifically allow out of bounds values (AFAIK only the "rate" Co for decks allows out of bounds values).

Be-ing commented:
The changes LGTM. Thanks for the code cleanup and making small commits that are easy to review. :)

Have you considered swapping the effect enabling and focusing? That is, having the focused effect selection be the default mode of the effect buttons and use Tap + effect buttons for enabling/disabling? When I played with the Pioneer DDJ-SB2, I found it more useful to have quick access to different effects' metaknobs than the enable switches. I generally don't manipulate the enable switches while I am using effects because that makes sudden changes in the sound, but having quick access to each metaknob allows for easier control of effect chains.

radusuciu committed:
Remove unnecessary soft takeover

radusuciu committed:
Remove unnecessary bounds checking

radusuciu committed:
Get/set sampler gain through parameter

radusuciu committed:
Remove redudant checking

radusuciu committed:
Switch mapping of effect focus and toggle

radusuciu commented:
I've addressed your review comments, and switched the mapping of effect toggle/focus. You're on point here: 

> I generally don't manipulate the enable switches while I am using effects because that makes sudden changes in the sound

I usually turn down the mix knob before making that change.

Be-ing commented:
Great, thanks for your work. The Mixtrack Pro 3 and Mixtrack 3 [are some of the best selling DJ controllers now](http://djtechtools.com/2017/02/07/what-were-the-top-10-selling-dj-controllers-in-2016/) (or at least they were last year), so having a good mapping for it will impact a lot of people. :) The VCI-400 mapping could also use updates for the new effects and looping UIs. Its effects controls will work well with the EffectUnit object in the Components library.

Have you tried the new reloop_andstop ControlObject? You can right click the reloop icon on screen to try it. I think having access to that is a big advantage for the way I mapped the 4 loop buttons on the DDJ-SB2.

radusuciu commented:
I think that the VCI-400 I have is different unfortunately as it's the Ean Golden edition which has different firmware. I'll see what I can do mapping wise. I do kinda want to keep it just for fun.. we'll see.

I haven't tried `reloop_andstop` but the current mapping fits my style well (or maybe it was the other way around) so I've been loathe to change it.

I'll take a final look at it and make sure nothing fucks up - will tag you back for a merge.

Be-ing commented:
> I think that the VCI-400 I have is different unfortunately as it's the Ean Golden edition which has different firmware. I'll see what I can do mapping wise.

Yeah it seems that firmware has lots of (apparently undocumented) differences from the regular VCI-400 firmware. In your JavaScript you may have to work around ugly hacks they did with the VCI-400 firmware to work around Traktor's limitations. It doesn't seem the firmware was ever made available to use with regular VCI-400s, so your mapping would only be usable by a select few people. :( But it looks like a nice controller.

Be-ing commented:
Yikes it looks like there was a miscommunication between Vestax and DJTT so that the firmware updates via USB of the VCI-400 and VCI-400 SE are incompatible...
http://forum.djtechtools.com/showthread.php?t=64071&p=572022&viewfull=1#post572022
http://forum.djtechtools.com/showthread.php?t=48174&page=20&p=478110&viewfull=1#post478110
https://serato.com/forum/discussion/942681

Be-ing commented:
Do you have any idea what is going on with the issue some users report on the forum of the right jog wheel not working?

radusuciu commented:
I don't know... Both my jog wheels definitely work. I think a clean install
of either 2.1 or 2.0 would rectify it barring any hardware issues. I
actually have had my jogs be unresponsive but that was while playing in a
dusty environment and it cleared up with a little tapping.

On Jun 25, 2017 11:29 PM, "Be" <notifications@github.com> wrote:

> Do you have any idea what is going on with the issue some users report on
> the forum of the right jog wheel not working?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/1287#issuecomment-310974095>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABDqeIEZs4VdnNiZeT8J5uWgBxJEdFl0ks5sH0-_gaJpZM4N_Rck>
> .
>

radusuciu commented:
Thanks for the links on the VCI-400SE, I've been doing some reading myself but only getting it today. Looks like I'll have my work cut out but it seems like it's achievable, if painful. One more off-topic question - do yo have any recs for MIDI debugging on Windows 10?

Back on the current topic, I think we're good to merge. Played with it last night and the latest changes behaved predictably. @daschuer? :)

daschuer commented:
Thank you for the work!
